controller/src/main/java/org/jboss/as/controller/AbstractOperationContext.java

```diff
@@ -1402,7 +1401,7 @@ class Step extends AbstractStep {
         private ResultHandler resultHandler;
         ServiceTarget serviceTarget;
         private ServiceVerificationHelper serviceVerificationHelper;
-        private Set<ServiceName> addedServices;
+        private Set<ServiceController<?>> addedServices;
```

Comment:
The backing set (created in getAddedServices()) is a HashSet, but ServiceControllerImpl does not implement hashCode semantics, nor does DelegatingServiceController delegate hashCode()/equals() to the underlying ServiceController.  That seems problematic.

Reply:
Good catch, thanks @pferraro ! It is fixed now.